### ISSUE TITLE
Cohort inclusion rule report - total events not matching records

### ISSUE THREAD

**Issue by anthonysena** [OHDSI/Atlas#1754]
### Expected behavior

After generating a cohort, viewing the report which displays the matches/events should match the event count.

### Actual behavior

This appears to only show the initial events and not the resulting records.

### Steps to reproduce behavior

Generated the following cohort on some of our data: [ami-cohort-example.txt](https://github.com/OHDSI/Atlas/files/3204154/ami-cohort-example.txt). 

Note the resulting display:
![image](https://user-images.githubusercontent.com/12902366/58126083-1f53ff00-7be0-11e9-9ed7-cc85ea91a348.png)

I would expect that the total events would match the records but perhaps I'm not fully clear on what these counts represent. Tagging @chrisknoll for clarification in case this is not a true bug.

**Comment by chrisknoll** [comment]
Records are cohort episodes.

Total Events are the candidate cohort entry events.  The Inclusion Report statistics are completely distinct from the final cohort statistics, and will not always line up.  If you could show us the 'by person' tab from the result of your screenshot above, you should see alignment in the 'people' column.

This is something that needs better documentation, but the top table is to show you what the cohort generation result is, wile the inclusion reports in the bottom give you information about what happened during generation.

**Comment by anthonysena** [comment]
Thanks @chrisknoll for clarifying. For the sake of completeness, here is the 'by person' tab from that same cohort and data source:

![image](https://user-images.githubusercontent.com/12902366/58173551-a8fbdf00-7c69-11e9-9815-92b2d75112f9.png)

And since the counts may be tough to read I'll list them here:

From the summary area:
- People: 710,097
- Records: 782,432

From the "By Person" tab
- Matches: 710,097
- Total Events: 710,097

From the "By Events" tab
- Matches: 1,797,397
- Total Events: 1,797,397

And for my own understanding (and with the intention to add this to the documentation), I'll attempt to draft the definitions for each of these here.

From the summary area:
- People == distinct count of people identified by the cohort definition
- Records == what amounts to the entries in the `cohort` table. That being the cohort_definition_id, subject_id (person_id), cohort_start_date, cohort_end_date. A single person can contribute > 1 record hence why the count may be higher with the cohort definition attached to this issue since it considers "all events" as specified in the definition.

The "By Person" tab:
- Matches == Total number of persons identified after applying all inclusion rules. In the cohort definition in this issue, there are no inclusion rules so...
- Total Events == Total number of people before applying all of the inclusion rules. 

The "By Events" tab:
- Matches == Total number of event records identified by the "cohort entry events". As you described, this is the candidate cohort entry events. 
- Total Events == Total number of event records after applying all of the inclusion rules.

I'm sure there are items above that need correction so please feel free to get out your red pen and edit!

**Comment by chrisknoll** [comment]
Only correction is the By Events:  In both cases (By Person and By Events) Total events represent the events selected by cohort entry, matching events represent those that meet all inclusion rules.  In your By Events tab, you have the descriptions swapped.

I like the way you are phrasing the 'By Person' description by making the description 'person focused'.  Matches is the total number of people after inclusion rules, and total events is the distinct person count from the set of cohort entry events.

However, it may be worthwhile to document the 'actual behavior', which in both cases, it is 'event-centric' but in the 'by person' view, the report picks one event per person, hence the events-to-person ratio is 1:1 (while for By Events, the ratio could be 1:N)

The rules for selecting the 'best event' is as follows:
1. For each person, pick the event that matches the most inclusion criteria.
2. In the event of a tie (ie: different events match the same number of inclusion criteria), choose the event that matches the lowest inclusion rule (ie, if you have an event that matches rule 1 ad another event that matches rule 2, the event selected for the 'by person' report will be the event that matches rule 1).

This might be a minor distinction from what you described being 'person centric', however when you have a case where there _are_ inclusion rules, it's important to know that the 'By Person' report is picking the 'best' event, so in my previous example (One person matching Inclusion 1 and Inclusion 2 at different events), the inclusion rule report will only show the person contributing counts to 'inclusion 1' and _not_ 'inclusion 2' because the single event selected was the event where the person matched inclusion 1.

This is an important thing to know, and I'll be sure to incorporate it into documentation.

**Comment by anthonysena** [comment]
Thanks @chrisknoll this is helpful and for the sake of posterity, I've updated my original definition to swap out the descriptions for the 'By Events' section based on your feedback. 

One more question: for the cohort entry events, you must define an initial event but you may also choose to define an "initial event restriction". When thinking about the "By Events" tab, I'm assuming that the events are those that meet the initial event criteria AND the initial event restriction. Do I have this right?

**Comment by chrisknoll** [comment]
>  I'm assuming that the events are those that meet the initial event criteria AND the initial event restriction. Do I have this right?

That is correct, Total Events are those events that are selected and subsequently filtered by the Initial Event Restriction (if initial event restriction is specified).  You could also thing of 'Total Events' as 'events that are evaluated by inclusion rules).  The Initial Event Restriction prevents events from being considered by the Inclusion Rules.

**Comment by chrisknoll** [comment]
This may make it more straight forward, let me know what you think:

- Total Events == The number of events selected from the Cohort Entry Events, after applying the Initial Event Restrictions, but before any Inclusion Rules were applied.

- Matches == The number of events from 'Total' that met all inclusion rules.

The Person and Events tab work off of different pools of events:

- The "By Person" tab:  For each person, a 'best' event is chosen based on the number of inclusion rules an event satisfies.  If there is a tie, then the event that matches the lowest inclusion rule is selected.  The end result of this is that only one event per person is selected, and therefore the numbers reported in the 'By Person' tab can be considered 'people'.

- The "By Events" tab: this tab uses all events that were applied to inclusion rules, so one person can have more than 1 event.  You can't look at the numbers in the 'By Events' tab as people _unless_ your cohort definition limited the initial events to _earliest_ or _latest_ per person.  In the case that the limit was specified, you will only have one event per person, and the By Person and By Event tabs will match.

### When to use the 'By Person' report
You are interested in knowing the 'best case scenario' of a population of people and those people matching all inclusion criteria, and you want to understand the number of people involved.

### When to use the 'By Events' report
You are interested in understanding how every event satisfies each inclusion rule.    People may satisfy one set of Inclusion Rules at one point, but a different set of inclusion rules at another time.  By looking across all events, you can understand how it was 'possible' to satisfy an inclusion criteria, but without knowing how many distinct people actually did (that's where you would go to the By Person report).

**Comment by anthonysena** [comment]
Thanks @chrisknoll for the additional description and for our offline discussion on this topic! I'll try to recap that discussion here for community benefit.

In reviewing the cohort described above in our environment, I was still curious why the total matched events (1,797,397) in the inclusion rule report still did not equal the total number of records in the summary (782,432). The discussion we had revealed the gap in my understanding: the inclusion rule report is focused on reporting the impact of the inclusion rules. The final number of records in the summary (782,432) is impacted by choices made in the **Cohort Exit** section and the **Cohort Eras** .

In the case of the AMI cohort, we designed it to use a fixed, 7-day duration persistence from index for the cohort exit. In addition, we defined a cohort era of 180d which will then combine AMI episodes into a single era if the occur in a 180d window from one another. When these choices are taken together with the total matched events, you move from 1,797,397 to 782,432 episodes and since we allow for all events, those 782,432 episodes are contributed by 710,097 people. If I were to change my cohort exit and/or cohort era, my final number of episodes would likely change.

Hoping I summarized everything we discussed and please feel free to add any additional feedback on this topic here so we can refer back to it for our documentation.

**Comment by pbr6cornell** [comment]
I understand what is being computed, but i dont think its reasonable to expect a user will fully understand, and moreover the statistics that they may expect to see are not available.  I do not recommend for v2.7.2, but lets revisit this as we plan for 3.0
